test/storage/bplustree_test.cpp

```diff
@@ -0,0 +1,1468 @@
+        bplustree->StructuralIntegrityVerification(*newkeys.begin(), *newkeys.rbegin(), &newkeys, bplustree->GetRoot()),
+        true);
+    EXPECT_EQ(newkeys.size(), 0);
+  }
+
+  // Insert Again
+  for (unsigned i = 0; i < 1000; i++) {
+    BPlusTree<int, TupleSlot>::KeyElementPair p1;
+    int k = rand_r(&globalseed) % 5000;
+    while (keys.find(k) != keys.end()) k++;
+    keys.insert(k);
+    p1.first = k;
+    bplustree->Insert(p1, predicate);
+
+    auto keys_copy = keys;
+
+    // Structural Integrity Verification Everytime
+    EXPECT_EQ(bplustree->StructuralIntegrityVerification(*keys_copy.begin(), *keys_copy.rbegin(), &keys_copy,
+                                                         bplustree->GetRoot()),
+              true);
+    EXPECT_EQ(keys_copy.size(), 0);
+  }
+
+  // Delete Again now two keys remaining
+  for (int i = 0; i < 999; i++) {
+    auto iter = keys.rbegin();
+    BPlusTree<int, TupleSlot>::KeyElementPair p1;
+    p1.first = *iter;
+    keys.erase(*iter);
+    EXPECT_EQ(bplustree->DeleteElement(p1), true);
+    std::set<int> newkeys = keys;
+
+    // Structural Integrity Test Everytime
+    EXPECT_EQ(
+        bplustree->StructuralIntegrityVerification(*newkeys.begin(), *newkeys.rbegin(), &newkeys, bplustree->GetRoot()),
+        true);
+    EXPECT_EQ(newkeys.size(), 0);
+  }
+
+  // Check Both still present
+  auto iter = keys.begin();
+  for (unsigned i = 0; i < 2; i++) {
+    EXPECT_EQ(bplustree->IsPresent(*iter), true);
+    iter++;
+  }
+
+  // Free Everything
+
+  delete bplustree;
+}
+
+void StructuralIntegrityTestWithRandomInsertAndDelete2() {
+  auto predicate = [](const TupleSlot slot) -> bool { return false; };
+
+  auto bplustree = new BPlusTree<int, TupleSlot>;
+  // The size is set to 2 more because of the following
+  // When we split an inner node, we might end up deleting an element
+  // from right side without putting anything in the right side
+  // Hence the size may be 31 at some points if we used 64.
+  bplustree->SetInnerNodeSizeUpperThreshold(6);
+  bplustree->SetLeafNodeSizeUpperThreshold(6);
+  bplustree->SetInnerNodeSizeLowerThreshold(2);
+  bplustree->SetLeafNodeSizeLowerThreshold(2);
+  std::set<int> keys;
+
+  for (unsigned i = 0; i < 37; i++) {
+    BPlusTree<int, TupleSlot>::KeyElementPair p1;
+    int k = rand_r(&globalseed) % 60;
+    while (keys.find(k) != keys.end()) k++;
+    keys.insert(k);
+    p1.first = k;
+    bplustree->Insert(p1, predicate);
+  }
+
+  auto it = keys.begin();
+  for (unsigned i = 0; i < keys.size() / 2; i++) {
+    BPlusTree<int, TupleSlot>::KeyElementPair p1;
+    p1.first = *(it);
+    it++;
+    EXPECT_EQ(bplustree->DeleteElement(p1), true);
+  }
+
+  for (unsigned i = keys.size() / 2; i < keys.size(); i++) {
+    BPlusTree<int, TupleSlot>::KeyElementPair p1;
+    p1.first = *(it);
+    it++;
+    EXPECT_EQ(bplustree->IsPresent(p1.first), true);
+  }
+
+  delete bplustree;
+}
+
+void StructuralIntegrityTestWithRandomInsertAndDelete2Reverse() {
```

Comment:
How is this different than the test above?

Reply:
This test deletes elements in the reverse order. Will include some meaningful documentation on what each test does while addressing @tpan496's comments below.